### PR TITLE
 [query] Refactor vds.sample_qc to use independent aggregators 

### PR DESCRIPTION
**Pull request by chrisvittal**
This will allow expert users to import these functions directly and use them in, for example, grouped aggregators.

```python3
from hail.vds.sample_qc import vmt_sample_qc, vmt_sample_qc_variant_annotations

vmt = vds.variant_data
vmt = vmt.annotate_entries(GT=hl.vds.lgt_to_gt(vmt.LGT, vmt.LA))
ac, atype = vmt_sample_qc_variant_annotations(global_gt=vmt.GT, alleles=vmt.alleles)
vmt = vmt.annotate_rows(group=compute_variant_group(vmt), ac=ac, atype=atype)
vmt = vmt.annotate_cols(
    qc=hl.agg.group_by(
        vmt.group,
        vmt_sample_qc(
            global_gt=vmt.GT,
            gq=vmt.GQ,
            ac=vmt.ac,
            atype=vmt.atype,
            dp=vmt.DP
        )
    )
)
```

### CODE REVIEW

**Comment by chrisvittal** [comment]
ci, test my changes!

File: hail/python/hail/vds/sample_qc.py [hunk 1]
```diff
@@ -0,0 +1,310 @@
...
+
+    require_first_key_field_locus(vds.reference_data, 'sample_qc')
+    require_first_key_field_locus(vds.variant_data, 'sample_qc')
+
+    if 'DP' in vds.reference_data.entry:
+        ref_dp_field_to_use = 'DP'
+    elif 'MIN_DP' in vds.reference_data.entry:
+        ref_dp_field_to_use = 'MIN_DP'
+    else:
+        ref_dp_field_to_use = dp_field
```
Comment (patrick-schultz) [inline hail/python/hail/vds/sample_qc.py:256]: I know this was the old behavior, but from the docstring I would expect `dp_field` to always be used if present. Should either reorder the branches here, or reword the docs.

File: hail/python/hail/vds/sample_qc.py [hunk 2]
```diff
@@ -0,0 +1,310 @@
...
+from hail.methods.misc import require_first_key_field_locus
+from hail.table import Table
+from hail.typecheck import sequenceof, typecheck, nullable
+from hail.utils.java import Env
+from hail.utils.misc import divide_null
+from hail.vds.variant_dataset import VariantDataset
+
+
+@typecheck(global_gt=Expression, alleles=ArrayExpression)
+def vmt_sample_qc_variant_annotations(
```
Comment (patrick-schultz) [inline hail/python/hail/vds/sample_qc.py:21]: For both expert users and ourselves, can you add some minimal documentation for these auxiliary methods? I don't know if they should be linked in the docs or not, but they should at least have docstrings.

File: hail/python/hail/vds/sample_qc.py [hunk 3]
```diff
@@ -0,0 +1,310 @@
...
+            locus=rmt.locus,
+            gq=rmt.GQ,
+            end=rmt.END,
+            dp=rmt_dp,
+            gq_bins=gq_bins,
+            dp_bins=dp_bins,
+        )
+    ).cols()
+
+    # TODO factor this out?
```
Comment (patrick-schultz) [inline hail/python/hail/vds/sample_qc.py:291]: I think it makes sense to factor out the combine variant and ref qc step. I'd expect anybody using the other factored out methods will need to do this.

**Review by patrick-schultz — DISMISSED** [review]
Very straightforward. A few comments, but otherwise looks good.

**Review by danking — DISMISSED** [review]
Can you expose `allele_ints` to users in some way? See the use-case here: https://hail.zulipchat.com/#narrow/stream/379853-Hail.2FVariants/topic/Echo.20VDS/near/423120704

**Review by patrick-schultz — APPROVED** [review]
Thanks!

**Comment by matren395** [comment]
Remarkable, thank you! Is this exposed in the most recent version of Hail , 0.2.128 ?

**Comment by patrick-schultz** [comment]
No, but it will be included in the next release, which should be in the next week or two.

**Comment by matren395** [comment]
Hmmmm we were looking to merge a PR featuring this code a bit before then. We've been working from a copy of this function I had just Copy-Pasted out into [some internal temp file called something like 'tmp_hail_methods'](https://github.com/broadinstitute/gnomad_qc/blob/5e96b3ef01f9d6ddaf80811c865a46cb7af04a15/gnomad_qc/v4/resources/temp_hail_methods.py) up until now, but I wouldn't want that temp file there when I merge. 

if it's not unreasonable would it be possible to get this in some 0.2.128.1 wheel ?

**Comment by chrisvittal** [comment]
No, it will be in 0.2.129 release date currently unknown, but we think it will be early March.

**Comment by matren395** [comment]
Hi!!!

- I was just delivered a wheel of Hail 0.2.128 loaded with this method from Chris, available at: gs://hail-30-day/hailctl/dataproc/cvittal-dev/0.2.128-2863e0eb192c/hail-0.2.128-py3-none-any.whl . Very helpful for removing those temp methods in the PR!

- it looks like the code will have a home inside of something that looks like the following: 
```
    qc = hl.struct(
        **{
            ann: hl.agg.filter(
                expr,
                vmt_sample_qc(
                    global_gt=vmt.GT,
                    gq=vmt.GQ,
                    variant_ac=vmt.variant_ac,
                    variant_atypes=vmt.variant_atypes,
                    dp=vmt.DP,
                ),
            )
            for ann, expr in argument_dictionary.items()
        }
    )
```
inside of a hl.agg.filter call, and where argument_dictionary looks like {'vep_name':ht.vep.field , "pass_all_vqc": hl.len(vmt.filters) == 0} and so on. Just to give an idea what the usage is going to be.

**Pull request merged**